File: module19,decision tree and random forest/assign3.R

# import dataset
fraud <- read.csv("C:/Users/USER/Downloads/Fraud_check.csv")
names(fraud)

# shifting taxable.income variable to front
col_idx <- grep("Taxable.Income", names(fraud))
Fraud <- fraud[, c(col_idx, (1:ncol(fraud))[-col_idx])]

# converting continous data into category 'risky and good'
fraud_check <- cut(Fraud$Taxable.Income, breaks =c(10000,30000,100000),labels=c("Risky","Good"),right=TRUE)
fraud_data <- cbind(Fraud,fraud_check)

# splitting data into training and testing

smp_size <- floor(0.75 * nrow(fraud_data))

## set the seed to make your partition reproducible
set.seed(123)
train_ind <- sample(seq_len(nrow(fraud_data)), size = smp_size)

train <- fraud_data[train_ind, ]
test <- fraud_data[-train_ind, ]

table(fraud_data$fraud_check)

# Step 3: Training a model on the data
install.packages("C50")
library(C50)
sales_model <- C5.0(train[ ,2:6], train[ ,7])
plot(sales_model)

train_pred <- predict(sales_model, train[ ,2:6])
train_acc <- mean(train[ ,7]==train_pred)
train_acc # 80.22%

# test accuracy
test_pred <- predict(sales_model, test[ ,2:6])
test_acc <- mean(test[ ,7]==test_pred)
test_acc # 76%

# cross tabulation of predicted versus actual classes
install.packages("gmodels")
library(gmodels)
CrossTable(test[ ,7], test_pred, prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
           dnn = c('actual default', 'predicted default'))

File: module19,decision tree and random forest/assign3.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 10 18:28:23 2020

@author: USER
"""

import pandas as pd
import numpy as np

Fraud = pd.read_csv("C:/Users/USER/Downloads/Fraud_check.csv")

# bringing target variable in front
cols = list(Fraud)
cols.insert(0, cols.pop(cols.index('Taxable.Income')))
cols
Fraud = Fraud.ix[:,cols]

# converting taxable.income into catgorical variable
x=pd.cut(Fraud["Taxable.Income"],bins=[0,30000,100000],labels=['Risky','Good'])
fraud=pd.concat([Fraud,x], axis=1)

from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()
fraud["Undergrad"]=lb.fit_transform(fraud["Undergrad"])
fraud["Urban"]=lb.fit_transform(fraud["Urban"])
fraud["Marital.Status"]=lb.fit_transform(fraud["Marital.Status"])

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train,test = train_test_split(fraud,test_size = 0.2)

X_train = np.array(train.ix[:,1:6]) # Input
X_test = np.array(test.ix[:,1:6]) # Input
Y_train = np.array(train.ix[:,6]) # Output
Y_test = np.array(test.ix[:,6]) # Output

from sklearn.tree import DecisionTreeClassifier as DT

help(DT)
model = DT(criterion = 'entropy')
model.fit(X_train,Y_train)

# Prediction on Train Data
preds = model.predict(X_train)
pd.crosstab(X_train,preds,rownames=['Actual'],colnames=['Predictions'])

np.mean(preds==Y_train) # 100% accuracy


# Prediction on Test Data
preds = model.predict(X_test)
pd.crosstab(Y_test,preds,rownames=['Actual'],colnames=['Predictions'])

np.mean(preds==Y_test) # 61.6% accuracy


















